.test-infra/mock-apis/src/main/go/internal/service/echo/echo.go

```diff
@@ -0,0 +1,185 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package echo contains the EchoService API implementation.
+package echo
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"path"
+	"reflect"
+	"time"
+
+	"github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/cache"
+	"github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/logging"
+	"github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/metric"
+	echov1 "github.com/apache/beam/test-infra/mock-apis/src/main/go/internal/proto/echo/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	metricsNamePrefix = "echo"
+	echoPath          = "/proto.echo.v1.EchoService/Echo"
+	echoPathAlias     = "/v1/echo"
+	healthPath        = "/grpc.health.v1.Health/Check"
+	healthPathAlias   = "/v1/healthz"
+)
+
+type Options struct {
+	Decrementer   cache.Decrementer
+	MetricsWriter metric.Writer
+	Logger        *slog.Logger
+	LoggingAttrs  []slog.Attr
+}
+
+// Register a grpc.Server with the echov1.EchoService. Returns a http.Handler or error.
+func Register(s *grpc.Server, opts *Options) (http.Handler, error) {
+	if opts.Logger == nil {
+		opts.Logger = logging.New(&logging.Options{
+			Name: reflect.TypeOf((*echo)(nil)).PkgPath(),
+		})
+	}
+	srv := &echo{
+		opts: opts,
+	}
+
+	echov1.RegisterEchoServiceServer(s, srv)
+	grpc_health_v1.RegisterHealthServer(s, srv)
+
+	return srv, nil
+}
+
+type echo struct {
+	echov1.UnimplementedEchoServiceServer
+	grpc_health_v1.UnimplementedHealthServer
+	opts *Options
+}
+
+// ServeHTTP implements http.Handler, allowing echo to support HTTP clients in addition to gRPC.
+func (srv *echo) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	switch r.URL.Path {
+	case echoPath, echoPathAlias:
+		srv.httpHandler(w, r)
+	case healthPath, healthPathAlias:
+		srv.checkHandler(w, r)
+	default:
+		http.Error(w, fmt.Sprintf("%s not found", r.URL.Path), http.StatusNotFound)
+	}
+}
+
+// Check checks whether echo service's underlying decrementer is alive.
+func (srv *echo) Check(ctx context.Context, _ *grpc_health_v1.HealthCheckRequest) (*grpc_health_v1.HealthCheckResponse, error) {
+	if err := srv.opts.Decrementer.Alive(ctx); err != nil {
+		return nil, err
+	}
+	return &grpc_health_v1.HealthCheckResponse{
+		Status: grpc_health_v1.HealthCheckResponse_SERVING,
+	}, nil
+}
+
+func (srv *echo) checkHandler(w http.ResponseWriter, r *http.Request) {
+	resp, err := srv.Check(r.Context(), nil)
+	if err != nil {
+
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		srv.opts.Logger.LogAttrs(context.Background(), slog.LevelError, err.Error(), srv.opts.LoggingAttrs...)
```

Comment:
Instead of using context.Background() for these logging calls, use the one attached to the request:

https://pkg.go.dev/net/http#Request.Context

Reply:
Ah this was a silly oversight on my part. I forgot there is a request.Context(). I will change this.